File: python/tests/test_informed_rrt_star.py
"""Informed RRT* (Gammell, Srinivasa & Barfoot 2014): reaches the goal on open
space, fails when walled off, and tightens toward the straight-line optimum."""

from __future__ import annotations

from pathlib import Path

import pytest
import yaml
from conftest import CONFIG_DIR, grid_from, open_grid

from navigation.core.params import ParamError, ParamSet
from navigation.global_planning import InformedRRTStar
from navigation.maps.occupancy_grid import OccupancyGrid2D

_ALGO = "informed_rrt_star"


def _capped_config(tmp_path: Path, max_iter: int) -> ParamSet:
    """Real config with a smaller iteration budget so the tests stay fast."""
    doc = yaml.safe_load((CONFIG_DIR / f"{_ALGO}.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] == "max_iterations":
            entry["default"] = max_iter
    out = tmp_path / f"{_ALGO}.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _motion_valid_path(grid: OccupancyGrid2D, path: list[tuple[float, float]]) -> bool:
    return all(grid.is_motion_valid(a, b) for a, b in zip(path, path[1:], strict=False))


def test_reaches_goal_on_open_grid(tmp_path: Path) -> None:
    grid = open_grid(10, 10, seed=7)
    start, goal = (0.5, 0.5), (9.0, 9.0)
    params = _capped_config(tmp_path, max_iter=2000)
    result = InformedRRTStar(params).plan(grid, start, goal)
    assert result.success
    assert result.path[0] == pytest.approx(start)
    assert grid.distance(result.path[-1], goal) <= params.get_float("goal_tolerance")
    assert _motion_valid_path(grid, result.path)
    assert result.cost > 0.0


def test_no_path_when_walled_off(tmp_path: Path) -> None:
    # Fully occupied middle column: no motion can cross it.
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."], seed=3)
    params = _capped_config(tmp_path, max_iter=400)
    result = InformedRRTStar(params).plan(grid, (0.5, 0.5), (4.5, 4.5))
    assert not result.success
    assert result.path == []
    assert result.cost == 0.0


def test_near_straight_line_optimality(tmp_path: Path) -> None:
    # On an obstacle-free grid, informed sampling should concentrate post-solution
    # draws in the ellipse and drive the cost close to the straight-line lower bound.
    grid = open_grid(12, 12, seed=5)
    start, goal = (0.5, 0.5), (11.0, 11.0)
    lower_bound = grid.distance(start, goal)
    result = InformedRRTStar(_capped_config(tmp_path, max_iter=4000)).plan(grid, start, goal)
    assert result.success
    assert result.cost <= lower_bound * 1.15


def test_out_of_range_default_rejected(tmp_path: Path) -> None:
    # goal_bias default above its declared max must fail validation at load time.
    doc = yaml.safe_load((CONFIG_DIR / f"{_ALGO}.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] == "goal_bias":
            entry["default"] = 2.0
    bad = tmp_path / f"{_ALGO}.yaml"
    bad.write_text(yaml.safe_dump(doc), encoding="utf-8")
    with pytest.raises(ParamError):
        ParamSet.from_yaml(bad)

File: cpp/tests/test_informed_rrt_star.cpp
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/core/params.hpp"
#include "navigation/global_planning/sampling/informed_rrt_star.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Point;

namespace {

// Open square grid (resolution 0.5): '.' everywhere.
std::vector<std::string> open_rows(int n) {
  return std::vector<std::string>(static_cast<size_t>(n), std::string(static_cast<size_t>(n), '.'));
}

// Generous anytime budget so informed sampling can tighten toward the optimum
// while the test stays fast.
std::string informed_cfg(int max_iterations) {
  return test::write_temp(
      "informed_rrt_star.yaml",
      "algorithm: informed_rrt_star\ncategory: global_planning\nparams:\n"
      "  - {name: max_iterations, type: int, default: " + std::to_string(max_iterations) +
          ", min: 1, max: 200000, description: n}\n"
      "  - {name: step_size, type: float, default: 0.5, min: 0.01, max: 100, description: eta}\n"
      "  - {name: goal_bias, type: float, default: 0.2, min: 0.0, max: 1.0, description: b}\n"
      "  - {name: goal_tolerance, type: float, default: 0.3, min: 0.0, max: 100, description: t}\n"
      "  - {name: neighbor_radius, type: float, default: 1.5, min: 0.01, max: 100, description: r}\n"
      "  - {name: radius_mode, type: enum, default: fixed, choices: [fixed, shrinking], description: m}\n"
      "  - {name: rgg_gamma, type: float, default: 2.0, min: 0.01, max: 100, description: g}\n"
      "  - {name: seed, type: int, default: 7, min: 0, max: 2147483647, description: s}\n");
}

}  // namespace

// (a) valid path, near the straight-line optimum on an open grid ---------------

TEST(InformedRrtStar, FindsNearOptimalPathOnOpenGrid) {
  auto g = test::make_grid(open_rows(12));  // 6m x 6m world
  Point start{0.25, 0.25}, goal{5.75, 5.75};
  global_planning::InformedRrtStarPlanner p(core::ParamSet::from_yaml(informed_cfg(8000)));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  // The last segment connects the tree to the exact goal.
  EXPECT_NEAR(r.path.back().x, goal.x, 1e-9);
  EXPECT_NEAR(r.path.back().y, goal.y, 1e-9);
  for (size_t i = 0; i + 1 < r.path.size(); ++i) {
    EXPECT_TRUE(g.is_motion_valid(r.path[i], r.path[i + 1])) << "path segment must be collision-free";
  }
  // Informed sampling should drive the cost close to the straight-line lower bound.
  double lower_bound = g.distance(start, goal);
  EXPECT_LE(r.cost, lower_bound * 1.15);
}

// (b) no-path case: a full wall separates start from goal ---------------------

TEST(InformedRrtStar, NoPathAcrossFullWall) {
  auto rows = open_rows(10);
  rows[5] = "##########";  // impassable horizontal wall
  auto g = test::make_grid(rows);
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::InformedRrtStarPlanner p(core::ParamSet::from_yaml(informed_cfg(400)));
  auto r = p.plan(g, start, goal, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
}

// (c) param validation failure: out-of-range default -------------------------

TEST(InformedRrtStar, BadGoalBiasThrows) {
  std::string bad = test::write_temp(
      "informed_rrt_star.yaml",
      "algorithm: informed_rrt_star\ncategory: global_planning\nparams:\n"
      "  - {name: goal_bias, type: float, default: 2.0, min: 0.0, max: 1.0, description: over max}\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}
